### PR TITLE
Add mechanism for integration tests

### PR DESCRIPTION
**Pull request by rth**
Partially addresses https://github.com/icubam/icubam/issues/301

From the added documentation,

> Integrations tests can be run with,
>
> ```
> pytest -m integration --icubam-config=<config.toml path>
> ```
> This will detect the database in the dev section of the config file,
> make copy it to a temporary location and provide a new config object
> with the `integration_config` pytest fixture. It can be used as follows,
>
> ```py
> @pytest.mark.integration
> def test_some_integration(integration_config):
>     db_path = integration_config.db.sqlite_path
>  ```
>
> If ``--icubam-config`` key is not provided, the above test would be skipped.
> If ``-m integration`` option is not provided, both unit and integration tests
> are run (see [pytest documentation](https://docs.pytest.org/en/latest/usage.html#specifying-tests-selecting-tests) for more details).

This would allow for instance to reproduce the https://github.com/icubam/icubam/issues/312 error (well after https://github.com/icubam/icubam/pull/222 is merged). Similarly it may help for https://github.com/icubam/icubam/issues/311 one could for instance get the correct AUTH token from the DB or the deployment hostname for production checks.

The actual DB passed via `--icubam-config` should never me modified, a backup is made, but just to be safe, maybe we should also open it in read-only mode.

Sorry this is all in pytest (https://github.com/icubam/icubam/issues/263), I don't know how to achieve this without it.

### CODE REVIEW

**Comment by dulacarnold** [comment]
Ok great, if this is an argument to move everything to pytest, *AND* we are confident we won't break the tornado tests in the process, I am for moving everything to pytest, we however need to commit to migrate to a situation where everything works, we shouldn't move half the tests to pytest and need to disable the tornado tests in the meantime or somesuch.

Should we also reconsider tox in this case? There wase some argument to try out tox and if we're going to migrate maybe that's worth the extra step?  Not at all arguing for this but just bringing it back on the table in case.

File: doc/contributing.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+### Integration tests
+
+Integrations tests can be run with,
+
+```
+pytest -m integration --icubam-config=<config.toml path>
+```
+This will detect the database in the dev section of the config file,
+make copy it to a temporary location and provide a new config object
```
Comment (dulacarnold) [inline doc/contributing.md:26]: "make a copy of it" or "copy it"?  or it's using Make to copy?

**Comment by dulacarnold** [comment]
Also, does this mean we should have a more realistically populated DB to test out things such as plotting?

**Comment by rth** [comment]
> Ok great, if this is an argument to move everything to pytest, AND we are confident we won't break the tornado tests in the process,

No it's not necessarily an argument for moving everything to pytest. The current situation is that we have classical unittests and unittests for tornado handlers, + some predicu tests that use pytest fixture. The whole things needs pytest to run  tests, but apart from that it doesn't require code migration one way or another. For tests in new modules, I personally would use plain test function (if there is not too much opposition from reviewers), but for existing ones they work as is and there is no need to change them. For simple changes, like https://github.com/icubam/icubam/pull/267 yes, we can migrate (or if someone wants to put in the effort), but I wouldn't migrate tornado handler tests, particularly as long as we have more pressing issues.

> Should we also reconsider tox in this case? There wase some argument to try out tox and if we're going to migrate maybe that's worth the extra step? 

I think this is orthogonal to pytest. tox basically reproduces part of CI locally, it's purely a dev tool. Personally I'm haven't seen the need to use it so for in this project, but if others developers find it useful we can certainly add it.

> Also, does this mean we should have a more realistically populated DB to test out things such as plotting?

Yes, this is intended to either test things locally on a real DB locally or before deployment, or we could try to make a real but heavily sanitized DB, say put it as a secret file in Github actions and run tests these tests on it in a separate job, only on upstream branches (not PRs). The questions is how to sync that integration DB with production. So maybe manual approach is still better.

**Pull request merged**